pinot-core/src/main/java/org/apache/pinot/core/data/manager/InstanceDataManager.java

```diff
@@ -86,11 +86,28 @@ void reloadSegment(String tableNameWithType, String segmentName)
       throws Exception;
 
   /**
-   * Reloads all segment in a table.
+   * Reloads all segments in a table.
    */
   void reloadAllSegments(String tableNameWithType)
       throws Exception;
 
+  /**
+   * Adds or replaces a segment in a table. Different from segment reloading,
+   * 1. this method doesn't assume the existence of TableDataManager object and it
+   * can actually initialize the TableDataManager for the segment;
+   * 2. this method can download a new segment to replace the local one before loading.
+   * Download is conducted when local segment's CRC is different from the one of the
+   * remote segment, but can also be forced to do regardless of CRC difference.
+   */
+  void addOrReplaceSegment(String tableNameWithType, String segmentName, boolean forceDownload)
```

Comment:
The `forceDownload` should not be part of this API. We can add separate APIs for `redownloadSegment` and `redownloadAllSegments`, or add the flag into the `reload` apis.

Reply:
as discussed, I'm adding the forceDownload flag to reload...() APIs, and refactor this addOrReplace...().